hudi-flink-datasource/hudi-flink1.18.x/src/main/java/org/apache/hudi/adapter/HoodieHiveCatalogAdapter.java

```diff
@@ -0,0 +1,138 @@
+import static org.apache.hudi.internal.schema.action.TableChange.ColumnPositionChange.ColumnPositionType.AFTER;
+import static org.apache.hudi.internal.schema.action.TableChange.ColumnPositionChange.ColumnPositionType.FIRST;
+import static org.apache.hudi.internal.schema.action.TableChange.ColumnPositionChange.ColumnPositionType.NO_OPERATION;
+
+/**
+ * Adapter clazz for {@code HoodieHiveCatalog}.
+ */
+public interface HoodieHiveCatalogAdapter extends Catalog {
+
+  default void alterTable(ObjectPath tablePath, CatalogBaseTable newCatalogTable, List<TableChange> tableChanges,
+      boolean ignoreIfNotExists) throws TableNotExistException, CatalogException {
+    checkNotNull(tablePath, "Table path cannot be null");
+    checkNotNull(newCatalogTable, "New catalog table cannot be null");
+
+    if (!isUpdateAllow(tablePath, newCatalogTable, ignoreIfNotExists)) {
+      return;
+    }
+    InternalSchema oldSchema = getInternalSchema(tablePath);
+    InternalSchema newSchema = oldSchema;
+    for (TableChange tableChange : tableChanges) {
+      newSchema = applyTableChange(newSchema, tableChange);
+    }
+    if (!oldSchema.equals(newSchema)) {
+      alterHoodieTableSchema(tablePath, newSchema);
+    }
+    refreshHMSTable(tablePath, newCatalogTable);
+  }
+
+  boolean isUpdateAllow(ObjectPath tablePath, CatalogBaseTable newCatalogTable, boolean ignoreIfNotExists) throws TableNotExistException;
+
+  InternalSchema getInternalSchema(ObjectPath tablePath) throws TableNotExistException, CatalogException;
+
+  void refreshHMSTable(ObjectPath tablePath, CatalogBaseTable newCatalogTable);
+
+  void alterHoodieTableSchema(ObjectPath tablePath, InternalSchema newSchema) throws TableNotExistException, CatalogException;
+
+  Type convertToInternalType(LogicalType logicalType);
+
+  default InternalSchema applyTableChange(InternalSchema oldSchema, TableChange change) {
+    InternalSchemaChangeApplier changeApplier = new InternalSchemaChangeApplier(oldSchema);
+    if (change instanceof TableChange.AddColumn) {
+      if (((TableChange.AddColumn) change).getColumn().isPhysical()) {
+        TableChange.AddColumn add = (TableChange.AddColumn) change;
+        Column column = add.getColumn();
+        String colName = column.getName();
+        Type colType = convertToInternalType(column.getDataType().getLogicalType());
+        String comment = column.getComment().orElse(null);
+        Pair<org.apache.hudi.internal.schema.action.TableChange.ColumnPositionChange.ColumnPositionType, String> colPositionPair =
+            parseColumnPosition(add.getPosition());
+        return changeApplier.applyAddChange(
+            colName, colType, comment, colPositionPair.getRight(), colPositionPair.getLeft());
+      } else {
+        throw new HoodieNotSupportedException("Add non-physical column is not supported yet.");
+      }
+    } else if (change instanceof TableChange.DropColumn) {
+      TableChange.DropColumn drop = (TableChange.DropColumn) change;
```

Comment:
can we use switch case clause?

Reply:
Switch not support class as case clause.
`DropColumn`, `AddColumn`... are subclass of `TableChange`, they are not using enum.